Add departure headings calculation

nataliejschultz opened the pull request:
In reference to #202:
Previously, generate_dataset would calculate the arrival heading using the preset osmnx `.bearing` method, while departure headings were set to `None`. 

This PR adds two functions written by @nreinicke:
1. `calculate_bearings`, which passes in linestrings and determines what to do with the coordinates based on the number of lat lon pairs
2. `compass_heading`, which calculates a compass bearing based on two lat lon pairs

These functions replace the osmnx method used previously, rather than using osmnx for arrival and the functions for departure. Both headings are added to the same data frame and output to edges-headings-enumerated.csv

nreinicke commented on python/nrel/routee/compass/io/generate_dataset.py:
```diff
@@ -209,3 +206,38 @@ def replace_id(vertex_uuid):
                 shutil.copy(model_path, model_dst)
+
+#Function written by Nick Reinicke
+def compass_heading(point1, point2):
```
Could we move these functions into the [utils](https://github.com/NREL/routee-compass/blob/main/python/nrel/routee/compass/io/utils.py) module?

nataliejschultz replied:
> Could we move these functions into the [utils](https://github.com/NREL/routee-compass/blob/main/python/nrel/routee/compass/io/utils.py) module?

Moved with commit b9f77261ebed959dd01386be240f4b2ee06d9537

nreinicke commented on python/nrel/routee/compass/io/generate_dataset.py:
```diff
@@ -209,3 +206,38 @@ def replace_id(vertex_uuid):
+
+    return compass_bearing
+
+#Function written by Nick Reinicke
```
I appreciate the attribution but I'm okay if we omit them 🙂

nreinicke reviewed (changes requested):
This is looking great! I left a couple of small comments but I like how simple you've kept this

nataliejschultz commented:
I added the two functions to `utils.py` and changed the relevant command:

`headings = [utils.calculate_bearings(i) for i in e.geometry.values]`

 For testing these changes locally, I added a name == main if statement to the bottom of the file with parameters to run:

```
if __name__ == '__main__':
    g = ox.graph_from_place("Denver, Colorado, USA", network_type="drive")
    testing_changes = generate_compass_dataset(g, 'testing-terminal')
```
Ran with `python ./routee-compass/python/nrel/routee/compass/io/generate_dataset.py` in terminal. After cd-ing into the testing-terminal output directory and using `gunzip`, here's what the top of the file looks like:

```
(base) nschultz-35728s:testing-terminal nschultz$ gunzip edges-headings-enumerated.csv.gz 
(base) nschultz-35728s:testing-terminal nschultz$ head edges-headings-enumerated.csv
arrival_heading,departure_heading
111,112
139,131
127,122
124,95
91,90
270,270
89,89
0,0
269,269
```
All is well!

nataliejschultz commented:
@nreinicke [tests failed ](https://github.com/NREL/routee-compass/actions/runs/10377622217) because the runner couldn't find utils, so I changed the `import utils` that worked on my machine to the more universal:

`from nrel.routee.compass.io import utils` 

in commit b6cbd623f0a46db11478fa61bc9262735dcd1658. Hopefully this solves it!

nreinicke reviewed (approved):
Excellent, looks good!
